File: components/cc1101custom/cc1101custom.py
import esphome.codegen as cg
import esphome.config_validation as cv

cc1101_ns = cg.esphome_ns.namespace("cc1101custom")
CC1101Custom = cc1101_ns.class_("CC1101Custom", cg.Component)

CONFIG_SCHEMA = cv.Schema({
    cv.GenerateID(): cv.declare_id(CC1101Custom),
    cv.Required("cs_pin"): cv.int_,
    cv.Required("gdo2_pin"): cv.int_,
})

async def to_code(config):
    var = cg.new_Pvariable(config["id"], config["cs_pin"], config["gdo2_pin"])
    await cg.register_component(var, config)

File: components/cc1101custom/cc1101custom.cpp
#include "cc1101custom.h"
#include "esphome/core/log.h"

namespace esphome {
namespace cc1101custom {

static const char *const TAG = "cc1101custom";

void CC1101Custom::setup() {
  ESP_LOGI(TAG, "CC1101 custom component initialized");
  ESP_LOGI(TAG, "CS pin: %d", this->cs_pin);
  ESP_LOGI(TAG, "GDO2 pin: %d", this->gdo2_pin);
}

}  // namespace cc1101custom
}  // namespace esphome
